exercises/practice/space-age/.approaches/introduction.md

```diff
@@ -0,0 +1,111 @@
+
+The various planets' orbital periods as compared to the Earth's are given.
+To convert them from _Earth years_ per orbit to _seconds_ per orbit you need to multiply this by the number of seconds in one Earth year.
+
+
+## Approach: give meaningful names to important values
+
+```haskell
+ageOn :: Planet -> Float -> Float
+ageOn planet seconds = seconds / (periodInEarthYears * secondsPerEarthYear)
+  where
+    secondsPerEarthYear = 60 * 60 * 24 * 365.25
+
+    periodInEarthYears = case planet of
+      Mercury -> 0.2408467
+      Venus -> 0.61519726
+      Earth -> 1
+      Mars -> 1.8808158
+      Jupiter -> 11.862615
+      Saturn -> 29.447498
+      Uranus -> 84.016846
+      Neptune -> 164.79132
+```
+
+This approach uses a `case` expression to choose the relevant orbital period.
+Also a `where` clause is used to give names to important values.
+This tends to greatly improve readability.
+
+
+## General guidance
+
+### `where` clauses are your friend!
+
+Giving meaningful names to subexpressions can do wonders for code readability.
+`let` expressions allow the same.
+Being expressions &ndash; which `where` clauses aren't &ndash; `let` expressions are a bit more flexible in their use.
+However, `where` clauses list the local definitions _after_ the main expression.
+This allows you to paint the broad strokes of your strategy first, and to fill in the details later.
+This is so convenient that it amply compensates for `where` clauses not being expressions.
+
+More on `where` and `let` elsewhere:
+
+- Haskell Wiki: [Let vs. Where][haskellwiki-let-vs-where]
+- Haskell Wikibook:
+  - [`where` clauses][wikibook-where]
+  - [`let` bindings][wikibook-let]
+  - [`let` and `where` revisited][wikibook-let-vs-where]
+
+
+### `case` expressions are also your friend!
+
+Many a beginning Haskeller writes code like
+
+```haskell
+ageOn planet seconds
+  | planet == Mercury = _
+  | planet == Venus   = _
+  | {- etc. -}        = _
+```
+
+This is an [anti-pattern][wikipedia-anti-pattern].
+Pattern-match instead, for example using a `case` expression:
+
+```haskell
+ageOn planet seconds =
+  case planet of
+    Mercury    -> _
+    Venus      -> _
+    {- etc. -} -> _
+```
+
+Pattern matching is a fundamental concept in Haskell, but this document is too short to be able to fully explain it.
+Please consult your other learning resources.
+
+Pattern matching with `case` has benefits over using guards:
+
+- `case` expressions are _expressions_.
+  Therefore as pieces of code they are very easy to move around during code composition.
+  You can give them names (e.g. in a `where` clause) and also pass them to functions as arguments.
+- The compiler is able to check that you handle all possible cases.
+  If you overlook some cases and use guards, the compiler will not help you.
+  But if you are pattern matching it will!
+- When you use pattern matching, the compiler can use its understanding of your code to apply code transformations that improve performance.
```

Comment:
> When you use pattern matching, the compiler can use its understanding of your code to apply code transformations that improve performance.

I suggest linking to another resource that explains the code transformation that it applies. As someone starting to learn Haskell, I curious to what and how Haskell do the code transformations.

Reply:
Yes, this would be nice. However, I know of no (_good_) such resources off the top of my head. I'll look for a bit.

Reply:
I haven't been able to find any and do not wish to delay merging for this, but would definitely welcome a PR in the future.